service/history/workflow/update/registry.go

```diff
@@ -181,17 +199,26 @@ func NewRegistry(
 				withInstrumentation(&r.instrumentation),
 			)
 		} else if acc := updInfo.GetAcceptance(); acc != nil {
+			if upd := r.updates[updID]; upd != nil {
+				_ = upd.advanceTo(stateAccepted)
+				return
+			}
 			r.updates[updID] = newAccepted(
 				updID,
 				acc.EventId,
 				r.remover(updID),
 				withInstrumentation(&r.instrumentation),
 			)
 		} else if updInfo.GetCompletion() != nil {
-			r.completedCount++
+			if upd := r.updates[updID]; upd != nil {
+				if err := upd.advanceTo(stateCompleted); err == nil {
+					upd.onComplete()
```

Comment:
nit: feel like something `advanceTo` should do when advancing to a terminal state.